zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -343,17 +343,48 @@ public void subscribeDataChanges(String path, IZkDataListener listener) {
    * Subscribe RecursivePersistListener for a particular path. User can only subscribe when
    * `_usePersistWatcher` is set to true and there is no pre-existing watcher on the path.
    */
-  // TODO: Add impl and remove exception
-  public boolean subscribePersistRecursiveWatcher(String path,
-      RecursivePersistListener recursivePersistListener)
-      throws KeeperException.UnimplementedException {
-    throw new KeeperException.UnimplementedException();
+  public void subscribePersistRecursiveListener(String path,
+      RecursivePersistListener recursivePersistListener) {
+    if (!_usePersistWatcher) {
+      throw new UnsupportedOperationException(
+          "Can not subscribe PersistRecursiveWatcher. Persist listener is not enabled.");
+    }
+
+    ManipulateListener addListener = () -> {
+      if (hasChildOrDataListeners(path)) {
+        throw new UnsupportedOperationException(
+            "Can not subscribe PersistRecursiveWatcher. There is an existing listener on " + path);
+      }
+      _zkPathRecursiveWatcherTrie.addRecursiveListener(path, recursivePersistListener);
+      getConnection().addWatch(path, ZkClient.this, AddWatchMode.PERSISTENT_RECURSIVE);
```

Comment:
Curious, what happens if we were not able to add watch on zk connection but update trie registry with listener. I'm guessing we do infinitely retry for addWatch operation but curious since "subscribePersistRecursiveListener" not an atomic operation now.

Reply:
Good point! Will add unsubscribe.

Reply:
After some thinking. I think it would be better to move `addRecursiveListener` after `getConnection().addWatch`. If addwatcher failed, it throws exception and the listener wont be added to trie.